File: android/src/main/java/com/mapconductor/react/geojson/GeoJSONStyleProviderRegistry.kt
package com.mapconductor.react.geojson

import android.content.Context
import com.mapconductor.geojson.GeoJSONStyleProviderInterface
import java.util.concurrent.ConcurrentHashMap

/** Application-owned native providers used by the React Native GeoJSON bridge. */
fun interface GeoJSONStyleProviderFactory {
    fun create(
        context: Context,
        sourceUri: String?,
    ): GeoJSONStyleProviderInterface
}

object GeoJSONStyleProviderRegistry {
    private val factories = ConcurrentHashMap<String, GeoJSONStyleProviderFactory>()

    fun register(
        id: String,
        factory: GeoJSONStyleProviderFactory,
    ) {
        require(id.isNotBlank()) { "GeoJSON style provider id must not be blank" }
        factories[id] = factory
    }

    fun create(
        id: String,
        context: Context,
        sourceUri: String?,
    ): GeoJSONStyleProviderInterface? = factories[id]?.create(context, sourceUri)
}

File: ios/GeoJSONStyleProviderRegistry.swift
import Foundation
import MapConductorGeoJSON

public typealias GeoJSONStyleProviderFactory = (
    _ sourceUri: String?
) throws -> any GeoJSONStyleProvider

/// Application-owned native style providers used by the React Native GeoJSON bridge.
public enum GeoJSONStyleProviderRegistry {
    private static let lock = NSLock()
    private static var factories: [String: GeoJSONStyleProviderFactory] = [:]

    public static func register(id: String, factory: @escaping GeoJSONStyleProviderFactory) {
        precondition(!id.isEmpty, "GeoJSON style provider id must not be empty")
        lock.lock()
        factories[id] = factory
        lock.unlock()
    }

    static func create(
        id: String,
        sourceUri: String?
    ) throws -> (any GeoJSONStyleProvider)? {
        lock.lock()
        let factory = factories[id]
        lock.unlock()
        return try factory?(sourceUri)
    }
}
